dex/ws/wslink.go

```diff
@@ -298,12 +289,39 @@ func (c *WSLink) outHandler(ctx context.Context) {
 		}
 	}
 
-	// On shutdown, relay ErrPeerDisconnected to any queued senders waiting on
-	// the outcome of the write (e.g. SendNow).
+	write := func(sd *sendData) {
+		c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+		err := c.conn.WriteMessage(websocket.TextMessage, sd.data)
+		if err != nil {
+			relayError(sd.ret, err)
+			// No more Sends should queue messages, and goroutines should return
+			// gracefully.
+			c.stop()
+			return
+		}
+		if sd.ret != nil {
+			close(sd.ret)
+		}
+	}
+
+	// On shutdown, process any queued senders before closing the connection, if
+	// it is still up.
 	defer func() {
-		log.Debugf("Shutting down link for %v with %v unsent messages.", c.ip, len(outQueue))
+		log.Infof("Shutting down link for %v with %v queued messages.", c.ip, len(outQueue))
+		// Drain the buffered channel of data sent prior to stop, but before it
+		// could be put in the outQueue.
+	out:
+		for {
+			select {
+			case sd := <-c.outChan:
+				outQueue = append(outQueue, sd)
+			default:
+				break out
+			}
+		}
+		// Attempt sending all queued outgoing messages.
 		for _, sd := range outQueue {
-			relayError(sd.ret, ErrPeerDisconnected)
+			write(sd)
```

Comment:
You are preserving the map now, but not in all cases. If a send error is encountered server-side before the client has reconnected, `removeClient` would already have been called, and we lose the `respHandlers`. Should we open an issue for this?

Reply:
Is this comment meant to be placed on the diff in https://github.com/decred/dcrdex/pull/406/commits/0ddc52f5dc6f6eb49b46db41a2da7c8e0f8b92dd?  I'm trying to discern if there's a connection between the two spots.

Although I do see that between `auth.connMtx.Unlock()` in the branch of `(*AuthManager).handleConnect` and the `auth.addClient` call to replace the clientInfo with the fresh connection that the map could be modified because it's a shallow copy.  Is that what you are referring to here?  I can make a deep copy of the map instead to resolve this.